pandas/tseries/tools.py

```diff
@@ -283,10 +297,23 @@ def to_datetime(arg, errors='raise', dayfirst=False, yearfirst=False,
     NaT
 
     """
-    return _to_datetime(arg, errors=errors, dayfirst=dayfirst,
-                        yearfirst=yearfirst,
-                        utc=utc, box=box, format=format, exact=exact,
-                        unit=unit, infer_datetime_format=infer_datetime_format)
+    if origin == 'unix':
+        return _to_datetime(arg, errors=errors, dayfirst=dayfirst, yearfirst=yearfirst,
+                            utc=utc, box=box, format=format, exact=exact,
+                            unit=unit, infer_datetime_format=infer_datetime_format)
+    elif origin == 'julian':
+        if unit != 'D':
+            raise AssertionError("unit must be 'D' for origin='julian'")
+        offset = tslib.Timestamp(0).to_julian_date()
+        arg = arg - offset
+        return _to_datetime(arg, errors=errors, dayfirst=dayfirst, yearfirst=yearfirst,
+                            utc=utc, box=box, format=format, exact=exact,
+                            unit=unit, infer_datetime_format=infer_datetime_format)
+    else:
+        offset = tslib.Timestamp(origin) - tslib.Timestamp(0)
+        return _to_datetime(arg, errors=errors, dayfirst=dayfirst, yearfirst=yearfirst,
+                            utc=utc, box=box, format=format, exact=exact,
+                            unit=unit, infer_datetime_format=infer_datetime_format) + offset
```

Comment:
instead of repeating all this code, just do:

```
offset = None
if origin == 'epoch':
    pass
elif origin == 'julian':
    # check errors
    arg = arg - Timestamp(0).to_julian_date()    
else:
    offset = Timestamp(origin) - Timestamp(0)

result = _to_datetime(.......)
if offset is not None:
     result = result + offset
return result
```